### ISSUE TITLE
[Meta] iOS/Android Background Execution

### ISSUE THREAD

**Issue by matthew-carroll** [flutter/flutter#32164]
iOS and Android have OS level restrictions about when things can run outside of a visible app, and how to run said code. This metabug tracks work items for bringing a reasonable support of background execution to Flutter.

In my opinion, what we should be aiming for when it comes to background execution is the ability to leverage platform-specific mechanisms via Flutter/Dart. Then, the Flutter team or external contributors can utilize those per-platform mechanics to build a cross-platform layer on top, i.e., generic background execution.

This metabug is based on conversation from this ticket:
- [x] https://github.com/flutter/flutter/issues/24278

**iOS**

- [x] [Offer a UIAppDelegate proxy for plugins](https://github.com/flutter/flutter/issues/32160)
- [x] [Offer a UNUserNotificationCenterDelegate proxy for plugins](https://github.com/flutter/flutter/issues/22099)
- [ ] [Background task execution](https://github.com/flutter/flutter/issues/32162)
- [ ] [OS managed downloads](https://github.com/flutter/flutter/issues/32161)

**Android**

- [ ] [Offer a Service that runs Dart code based on Intent arguments](https://github.com/flutter/flutter/issues/22234)
- [ ] [Offer a WorkManager task that runs Dart code](https://github.com/flutter/flutter/issues/22234)

**Dart**

- [ ] [Add a process-wide plugin isolate](https://github.com/flutter/flutter/issues/23904)

**Cross-Platform**

Once the per-platform behaviors are well implemented, we should evaluate what might make sense as cross-platform background execution.

**Comment by matthew-carroll** [comment]
CC @Hixie @dnfield @mklim @bkonyi @xster @LongCatIsLooong @tvolkert @mehmetf

**Comment by mikeryder** [comment]
iOS supports long-running tasks for specific purposes (e.g. streaming music, location updates, Bluetooth communication, etc.).  I think support for those should be included as part of the "iOS: Background task execution" or a new bullet under iOS.

**Comment by matthew-carroll** [comment]
@mikeryder do those things actually require dedicated attention? When I read through the iOS docs on background execution, it looked like each of those use-cases can be enabled by simply updating the info.plist. Can't developers write plugins for those behaviors right now?

**Comment by MaikuB** [comment]
should these two be added to the new meta-bug as well?

https://github.com/flutter/flutter/issues/22099
https://github.com/flutter/flutter/issues/23904

the first one did linked in the original issue that's been closed. the second one didn't but is related to background execution

**Comment by mikeryder** [comment]
@matthew-carroll Yes, your're right, long-running tasks can be implemented by writing a plugin, but the implementation would be native.  I thought the goal of this meta-bug was to allow Dart code to be run in the background in the variety of ways each native platform supports.  That is the piece that is not currently supported.

**Comment by matthew-carroll** [comment]
Thanks @MaikuB I added both of those issues to the ticket.

**Comment by matthew-carroll** [comment]
@mikeryder the native pieces are what allow Dart code to run in the background. Maybe I'm misunderstanding your use-cases. Can you describe the one or more specific goals that you have personally for Dart background execution?

**Comment by audkar** [comment]
Our use case was:
Connect to Bluetooth device. Collect data from that BL device, GPS location and some sensors for long period of time (until manually disconnect).

To accomplish this it is required to have Foreground service on Android. We expected that we could use existing plugins to collect all data (geolocation, ble and sensors plugins). However we failed to make this work in one week and decided to make native apps.

Main problem was that this task was long running. Foreground service is alive for a long period of time and during that period Activity is destroyed on devices. Sometimes Foreground service can be started from BroadcastReceiver and Activity is not started at all.

**Comment by matthew-carroll** [comment]
@audkar you mentioned that you tried using existing plugins. Is there any reason that the foreground Service capability should be a part of the core framework rather than part of the plugins that need to run as a foreground Service? In other words, if those plugins had allowed you to use them as a foreground Service, would that have solved your problems?

**Comment by audkar** [comment]
Problem was that there had be one Foreground service shared between all three plugins. 

Yes all plugins can create its own Foreground services. But then would be 3 notifications indicating that three services are running.

**Comment by matthew-carroll** [comment]
@audkar that makes sense. I actually already have a ticket related to foreground Services. I'll add it to this metabug.

**Comment by a7me63azzab** [comment]
Any updates ? Currently , I'm working on a project that needs to calculate location and user steps in background ??

**Comment by matthew-carroll** [comment]
Hi @a7me63azzab, can you let me know which of the individual tickets in this metabug are required for you to accomplish your goals?

**Comment by audkar** [comment]
@matthew-carroll I have created oversimplified example [application](https://github.com/audkar/native-android-sensors) to represent problem we faced. Managed to make it under 2h. We spend almost one full week trying to do same thing with Flutter (to Flutter defense we had no previous experience with it :) platform-channels, isolates, plugins, etc). (and failed). Problems we faced I have listed in comments above :point_up:

To be on the same page and avoid any miscommunication problems I created this sample project. Do you think it is possible to achieve same result with current Flutter APIs?

**Comment by a7me63azzab** [comment]
Hi @matthew-carroll 
My use case was:
I want to collect data from GPS location and Pedometer sensor even if app is closed,
So after searching , I found two packages that i can use :
1- geolocator 
2- pedometer
 but these plugins working only when app is running .
 so i tried to run these plugins in isolates but i get an error.
```
I/flutter (21632): ══╡ EXCEPTION CAUGHT BY SERVICES LIBRARY ╞══════════════════════════════════════════════════════════
I/flutter (21632): The following _CompileTimeError was thrown while activating platform stream on channel
I/flutter (21632): pedometer.eventChannel:
I/flutter (21632): error: native function 'Window_sendPlatformMessage' (4 arguments) cannot be found
I/flutter (21632): When the exception was thrown, this was the stack:
I/flutter (21632): #0      Window.sendPlatformMessage (dart:ui/window.dart:868:9)
I/flutter (21632): #1      BinaryMessages._sendPlatformMessage (package:flutter/src/services/platform_messages.dart:46:15)
I/flutter (21632): #2      BinaryMessages.send (package:flutter/src/services/platform_messages.dart:97:12)
I/flutter (21632): #3      MethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart:295:50)
I/flutter (21632): <asynchronous suspension>
I/flutter (21632): #4      EventChannel.receiveBroadcastStream.<anonymous closure> (package:flutter/src/services/platform_channel.dart:490:29)
I/flutter (21632): <asynchronous suspension>
I/flutter (21632): #12     StepCounterState.initPlatformState (package:stepcounter/main.dart:59:30)
I/flutter (21632): <asynchronous suspension>
I/flutter (21632): #13     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:303:17)
I/flutter (21632): #14     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:171:12)
I/flutter (21632): (elided 7 frames from package dart:async)

```
And i found that plugins only can be used in the main UI thread.
After that we decided to make native packages to collect data from GPS location and Pedometer sensor and save them in sqlite file , and then read data from that file using flutter app.
So, I just asking if there is any update in flutter background execution to avoid playing with native platforms ???

**Comment by fre391** [comment]
Hi, I like the idea of a local sqlite for datasharing.... could you pls keep me updated about this idea..thx

Did you notice this one?
https://github.com/Lyokone/flutterlocation/wiki/Background-Location-Updates

**Comment by matthew-carroll** [comment]
@a7me63azzab when you tried using the plugins in the background, did you make sure to invoke `FlutterMain.startInitialization()` and `FlutterMain.ensureInitializationComplete()` from your Android `Service`? Also, from the Dart side, did you make sure to invoke `runApp()` from your entrypoint? I'm curious if one of those details is the reason for your error.

**Comment by a7me63azzab** [comment]
Hi, @matthew-carroll   thanks for your response , 
>did you make sure to invoke ```runApp()``` from your entrypoint? 
```
void main() {
  runApp(StepCounter());
}
```
>did you make sure to invoke ``` FlutterMain.startInitialization() ``` and ```FlutterMain.ensureInitializationComplete()``` from your Android Service?
 
After adding ``` FlutterMain.startInitialization() ``` and ```FlutterMain.ensureInitializationComplete()``` in MainActivity.java i get an error , Is MainActivity.java what you mean by android Service or what ?
```
package com.example.stepcounter;

import android.os.Bundle;
import io.flutter.app.FlutterActivity;
import io.flutter.plugins.GeneratedPluginRegistrant;

public class MainActivity extends FlutterActivity {
  @Override
  protected void onCreate(Bundle savedInstanceState) {
    FlutterMain.startInitialization(this);
    FlutterMain.ensureInitializationComplete(this);
    super.onCreate(savedInstanceState);
    GeneratedPluginRegistrant.registerWith(this);
  }
}

```

**Comment by matthew-carroll** [comment]
I meant an Android Service: https://developer.android.com/reference/android/app/Service

If you're attempting to run code in the background, it should be running in a Service, not an Activity.

**Comment by ryanheise** [comment]
In addition to exposing the `IntentService` and `WorkManager` APIs to Dart, I would point out that there are numerous specialised APIs here, including `MediaBrowserServiceCompat` for background audio, and so this could be a big project, and hopefully there is some prioritisation.

So one thing that I would prioritise is simply improved documentation on the current v2 APIs. The current [plugin migration guide](https://flutter.dev/docs/development/packages-and-plugins/plugin-api-migration) from v1 to v2 does not touch on background execution and does not explain (or link to explanations of) the underlying model and concepts behind v2.

Another priority could be to update the Medium article on Flutter background execution post v2, just so plugin authors can quickly help users through this transition. These things are faster to do than the more ambitious items on the todo list and would have immediate benefit.

**Comment by realsoc** [comment]
Any updates?
My use cases would be 
1. a countdown timer to cook pasta with a time range in 1seconds - 24hours that rings even if the screen is off or the app background.
2. an application that detects a fall when riding a bike through the sensors. Same thing here, the app could be background or the screen off.

**Comment by vkammerer** [comment]
> Problem was that there had be one Foreground service shared between all three plugins.
> 
> Yes all plugins can create its own Foreground services. But then would be 3 notifications indicating that three services are running.

I am facing a similar issue in my own application, in which I want a Foreground Service to execute Dart callbacks triggered by several plugins. I found a way to solve it, and generalised it to accept more use cases, so I'm sharing it here it the hope that it might help others too.

Also, this approach has served me well so far, but I am not an experienced mobile developer, nor familiar with the inner workings of Flutter, so I'd be very happy to hear what the shortcomings of such an approach could be.

My solution was to:
1. create a package called [flutter_service](https://github.com/vkammerer/flutter_service), which implements:
- an adaptation of the [`FlutterBackgroundExecutor`](https://github.com/flutter/plugins/blob/master/packages/android_alarm_manager/android/src/main/java/io/flutter/plugins/androidalarmmanager/FlutterBackgroundExecutor.java) class of the `android_alarm_manager` plugin
- a general purpose `JobIntentService`
- a general purpose Foreground Service

2. Adapt existing plugins so that they depend on `flutter_service`:
- [`flutter_local_notifications_broadcast`](https://github.com/vkammerer/flutter_local_notifications_broadcast) broadcasts an Intent that starts the Foreground Service
- [`android_alarm_manager`](https://github.com/vkammerer/android_alarm_manager) broadcasts an Intent that executes a callback either in the `JobIntentService` or the Foreground Service

The workflow for my app can be represented in this diagram:   

![example diagram for flutter_service](https://vkammerer.github.io/flutter_service/images/flutter_service_example.svg?)

A more detailed explanation of what `flutter_service` does can be found on its [GH repository](https://github.com/vkammerer/flutter_service).

What do you think about this design?

**Comment by ryanheise** [comment]
Hi @vkammerer you can also look at my [audio_service](https://pub.dev/packages/audio_service) plugin which I created back in late 2018. It takes a similar approach to what you described, but implements `MediaBrowserServiceCompat` rather than a generic foreground service, and uses the other various platform APIs that are supposed to go along with `MediaBrowserServiceCompat` like `MediaSessionCompat` and the various APIs for handling audio focus, wake locks, etc.

In the bigger picture, it would be good to have a separate plugin like this for WorkManager, and so on, for each other specialised kind of use case.

Of course, I think this won't completely eliminate the need for plugins that need to build their own foreground service, or WorkManager, because there will always be certain highly specialised use cases where, for example, a custom view needs to be set on the notification, or there needs to be some complex interplay between other platform APIs and the service for it to work.

But I think these sorts of general plugins will still be really useful in many common cases.

**Comment by karmazinkd** [comment]
@vkammerer how does it behave on iOS? 
As far as I know, there are system limitation of the background fetch intervals, starting with once per 15 minutes and up to days before any wake is ermitted (iOS decides based on your activity when to permit apps to fetch in the background)

**Comment by AAverin** [comment]
There are many different specific cases like playing the audio, bluetooth handling, location and sensors.
Still, there is a very generic case that is still needs to be handled – a long running background operation with active foreground notification.
I operate in Android terms and not sure if there is an exact equivalent on iOS.

Basic idea is to be able to build some kind of SyncService – a service that could do some long-running operations to sync with the backend, talking to APIs, downloading files, saving them to local storage, doing some heavy operations like archive/unarchive and so on. Of course, user wouldn't wait for this to finish and there should be some way to show the progress.

**Comment by Noyon01631** [comment]
Good

**Comment by ryanheise** [comment]
There are very specific use cases on iOS where background execution is permitted, and so you unfortunately can't run code in the background on iOS unless you have one of the approved purposes (playing music being one of them). In practice, for plugins that provide a unified API across iOS and Android, we will need to have specialised plugins for each typical background use case that is recognised by the platforms. On a side note, the main reason why iOS supports background execution for particular use cases, and why Android at least in recent years has moved in that direction, is to ensure that battery usage is managed optimally for those different use cases.

So basically to handle background execution correctly we will ultimately have a separate plugin for background fetch, for background audio, etc. etc.

**Comment by AAverin** [comment]
On Android we should be able to run custom code using bound services and foreground notifications.
On iOS the closest equivalent seems to be https://developer.apple.com/documentation/backgroundtasks. At least it should let to run minutes long operations.

**Comment by bhanuka96** [comment]
> Any updates ? Currently , I'm working on a project that needs to calculate location and user steps in background ??

Did ou find solution for both Android and Ios?

**Comment by tank777** [comment]
Is there any progress or update on this?

**Comment by stuartmorgan** [comment]
Given the limitations on background execution on iOS, as discussed above, creating a general-purpose is likely not viable. There are community plugins for at least some specific use cases already, such as https://pub.dev/packages/workmanager.

**Comment by lionuncle** [comment]
any ETA? ( June 24, 2021)

**Comment by ScottishRoss** [comment]
Commenting to keep track of the progress of this.

**Comment by Levi-Lesches** [comment]
@ScottishRoss, you can just press "Subscribe" on the left side to get notifications without sending everyone else one :)

**Comment by shyney7** [comment]
https://docs.flutter.dev/development/packages-and-plugins/background-processes

**Comment by tudor07** [comment]
@shyney7 that documentation contains two useless paragraphs and links to a 4 year old article.
Flutter really needs a serious background process framework.

**Comment by tc14077** [comment]
@tudor07 the way that introduced in the article has been updated in last year (09/10/2020), and the update reflects the move to the v2 Android embedding API. Therefore, the article should be valid at this point. 

**_But_**, I agree with your view. Flutter really needs a serious and **_intuitive enough_** background process framework.

**Comment by ljfreelancer88** [comment]
My use case is play/pause YouTube Iframe using API in the background. Something like putting the foreground UI in the background when the device is in lockscreen mode.

**Comment by DesmondFox** [comment]
Any updates?

**Comment by manuelduarte077** [comment]
I think we are still with the same problem.

I'm looking for how to use geofencing

**Comment by ryanheise** [comment]
See the [official response](https://github.com/flutter/flutter/issues/32164#issuecomment-829473241) above.

We have plugins to handle specific background use cases:

* [audio_service](https://pub.dev/packages/audio_service)
* [flutter_background_geolocation](https://pub.dev/packages/flutter_background_geolocation)
* [flutter_downloader](https://pub.dev/packages/flutter_downloader)
* [workmanager](https://pub.dev/packages/workmanager)
* [flutter_background_service](https://pub.dev/packages/flutter_background_service)

To name a few.

**Comment by manuelduarte077** [comment]
yes, I was seeing other uses, which were on the same topic and had the same response

**Comment by tudor07** [comment]
@ryanheise those are not enough/not working right. I've given up hope Flutter will ever fix this.

**Comment by manuelduarte077** [comment]
> > He perdido la esperanza de que Flutter alguna vez arregle esto.
> 
> Por supuesto. Flutter ya indicó hace más de un año que es poco probable que aborde esto. Vuelva a ver la [respuesta oficial](https://github.com/flutter/flutter/issues/32164#issuecomment-829473241) de abril de 2021 donde se asignó la prioridad P6 (lo que indica que es poco probable que se aborde) y se sugirió Workmanager como complemento de ejemplo.
> 
> > esos no son suficientes/no funcionan bien.
> 
> Esa respuesta oficial también sugirió workmanager.
> 
> Veo en su actividad de GitHub que tuvo problemas con background_fetch, por lo que tal vez eso fue lo que insinuó como "no funciona bien". Pero, ¿ha probado esa sugerencia del administrador de trabajo de hace un año? Ya proporciona asignaciones para `requiresExternalPower`( `requiresCharging`) y `requiresNetworkConnectivity`( `networkType`) que parece querer para su caso de uso, de acuerdo con su actividad de GitHub.

Those resources that spend there, have already been published for quite some time, and also with examples that are not very clear in the end I had to guide myself from an example that has a good time, but I do think that this one in flutter is something very cumbersome and you have to do many laps to achieve it

**Comment by stuartmorgan** [comment]
> Given the limitations on background execution on iOS, as discussed above, creating a general-purpose plugin is likely not viable. There are community plugins for at least some specific use cases already, such as https://pub.dev/packages/workmanager.

Per the updated guidance in https://github.com/flutter/flutter/wiki/Issue-hygiene#closing-issues, closing, to better reflect that this is not something the Flutter team has plans to implement, as community plugins can address specific use cases that are possible to support.

**Comment by AAverin** [comment]
https://pub.dev/packages/workmanager doesn't have scheduling support on iOS and has no support for MacOS.
So if anyone following this issue has the capacity, support and PRs are warmly welcome.
